pkg/cmd/manager/controller.go

```diff
@@ -0,0 +1,310 @@
+func (bm BaseManager) GetWatchNamespace() (string, error) {
+	ns, found := os.LookupEnv(bm.WatchNamespaceEnvVar)
+	if !found {
+		return "", fmt.Errorf("%s must be set", bm.WatchNamespaceEnvVar)
+	}
+	return ns, nil
+}
+
+func (bm BaseManager) CreateBootstrapClient(cfg *rest.Config) (client.Client, error, string) {
+	// Increase maximum burst that is used by client-side throttling,
+	// to prevent the requests made to apply the bundled Kamelets
+	// from being throttled.
+	cfg.QPS = 20
+	cfg.Burst = 200
+	bootstrapClient, err := client.NewClientWithConfig(false, cfg)
+	if err != nil {
+		return nil, err, "cannot initialize client"
+	}
+
+	return bootstrapClient, nil, ""
+}
+
+func (bm BaseManager) CreateEventBroadcaster(bootstrapClient client.Client, watchNamespace string, ctx context.Context) (record.EventBroadcaster, error, string) {
+	// We do not rely on the event broadcaster managed by controller runtime,
+	// so that we can check the operator has been granted permission to create
+	// Events. This is required for the operator to be installable by standard
+	// admin users, that are not granted create permission on Events by default.
+	broadcaster := record.NewBroadcaster()
+
+	if ok, err := kubernetes.CheckPermission(ctx, bootstrapClient, corev1.GroupName, "events", watchNamespace, "", "create"); err != nil || !ok {
+		// Do not sink Events to the server as they'll be rejected
+		broadcaster = event.NewSinkLessBroadcaster(broadcaster)
+		return broadcaster, err, "cannot check permissions for creating Events"
+		bm.Log.Info("Event broadcasting is disabled because of missing permissions to create Events")
+	}
+
+	return broadcaster, nil, ""
+}
+
+func (bm BaseManager) GetControllerNamespaceAndLeaderElection(watchNamespace string, bootstrapClient client.Client, leaderElection bool, ctx context.Context) (string, bool, error, string) {
+	operatorNamespace := platform.GetOperatorNamespace()
+	if operatorNamespace == "" {
+		// Fallback to using the watch namespace when the operator is not in-cluster.
+		// It does not support local (off-cluster) operator watching resources globally,
+		// in which case it's not possible to determine a namespace.
+		operatorNamespace = watchNamespace
+		if operatorNamespace == "" {
+			leaderElection = false
+			bm.Log.Info("unable to determine namespace for leader election")
+		}
+	}
+
+	if ok, err := kubernetes.CheckPermission(ctx, bootstrapClient, coordination.GroupName, "leases", operatorNamespace, "", "create"); err != nil || !ok {
+		leaderElection = false
+		if err != nil {
+			return operatorNamespace, leaderElection, err, "cannot check permissions for creating Leases"
+		}
+		bm.Log.Info("The operator is not granted permissions to create Leases")
+	}
+
+	return operatorNamespace, leaderElection, nil, ""
+}
+
+// SetOperatorImage set the operator container image if it runs in-container
+func (bm BaseManager) SetOperatorImage(bootstrapClient client.Client, ctx context.Context) (error, string) {
```

Comment:
it seems this function does not really belong to the `Manager` interface as in fact it sets the image elsewhere. It may be moved to a simple function so it may simplify the  `Manager` interface.